src/x86simdsort-static-incl.h

```diff
@@ -0,0 +1,148 @@
+    { \
+        ISA##_qsort(arr, size, hasnan, descending); \
+    } \
+    template <typename T> \
+    X86_SIMD_SORT_FINLINE void x86simdsortStatic::qselect( \
+            T *arr, size_t k, size_t size, bool hasnan, bool descending) \
+    { \
+        ISA##_qselect(arr, k, size, hasnan, descending); \
+    } \
+    template <typename T> \
+    X86_SIMD_SORT_FINLINE void x86simdsortStatic::partial_qsort( \
+            T *arr, size_t k, size_t size, bool hasnan, bool descending) \
+    { \
+        ISA##_partial_qsort(arr, k, size, hasnan, descending); \
+    } \
+    template <typename T> \
+    X86_SIMD_SORT_FINLINE void x86simdsortStatic::argsort( \
+            T *arr, size_t *arg, size_t size, bool hasnan, bool descending) \
+    { \
+        ISA##_argsort(arr, arg, size, hasnan, descending); \
+    } \
+    template <typename T> \
+    X86_SIMD_SORT_FINLINE std::vector<size_t> x86simdsortStatic::argsort( \
+            T *arr, size_t size, bool hasnan, bool descending) \
+    { \
+        std::vector<size_t> indices(size); \
+        std::iota(indices.begin(), indices.end(), 0); \
+        x86simdsortStatic::argsort( \
+                arr, indices.data(), size, hasnan, descending); \
+        return indices; \
+    } \
+    template <typename T> \
+    X86_SIMD_SORT_FINLINE void x86simdsortStatic::argselect( \
+            T *arr, size_t *arg, size_t k, size_t size, bool hasnan) \
+    { \
+        ISA##_argselect(arr, arg, k, size, hasnan); \
+    } \
+    template <typename T> \
+    X86_SIMD_SORT_FINLINE std::vector<size_t> x86simdsortStatic::argselect( \
+            T *arr, size_t k, size_t size, bool hasnan) \
+    { \
+        std::vector<size_t> indices(size); \
+        std::iota(indices.begin(), indices.end(), 0); \
+        x86simdsortStatic::argselect(arr, indices.data(), k, size, hasnan); \
+        return indices; \
+    }
+
+/*
+ * qsort, qselect, partial, argsort key-value sort template functions.
+ */
+#include "xss-common-qsort.h"
+#include "xss-common-argsort.h"
+
+#if defined(__AVX512DQ__) && defined(__AVX512VL__)
+/* 32-bit and 64-bit dtypes vector definitions on SKX */
+#include "avx512-32bit-qsort.hpp"
+#include "avx512-64bit-qsort.hpp"
+#include "avx512-64bit-argsort.hpp"
+#include "avx512-64bit-keyvaluesort.hpp"
+
+/* 16-bit dtypes vector definitions on ICL */
+#if defined(__AVX512BW__) && defined(__AVX512VBMI2__)
+#include "avx512-16bit-qsort.hpp"
+/* _Float16 vector definition on SPR*/
+#if defined(__FLT16_MAX__) && defined(__AVX512BW__) && defined(__AVX512FP16__)
+#include "avx512fp16-16bit-qsort.hpp"
+#endif // __FLT16_MAX__
+#endif // __AVX512VBMI2__
+
+XSS_METHODS(avx512)
+
+// key-value currently only on avx512
+template <typename T1, typename T2>
+X86_SIMD_SORT_FINLINE void
+x86simdsortStatic::keyvalue_qsort(T1 *key, T2 *val, size_t size, bool hasnan)
+{
+    avx512_qsort_kv(key, val, size, hasnan);
+}
+
+#elif defined(__AVX2__) && !defined(__AVX512F__)
```

Comment:
So, this will fail to compile for a few configurations that do support AVX2 and AVX512F but not AVX512DQ and VL (for example, -march=knl). Up to you whether to leave it that way, or change it so it would build the AVX2 version

Reply:
yeah that is deliberate. Someone compiling with -mavx512f shouldn't get the avx2 version without being warned about it.

Reply:
Okay that makes sense. It might be nice to make the error message more specific in that case, maybe mentioning that AVX512DQ and AVX512VL are needed if AVX512F is enabled but nothing else.

Reply:
Yeah makes sense. Perhaps we could do that and update README in a separate PR.